### DISCUSSION TITLE
Эффекты заклинаний

### DISCUSSION THREAD

**Discussion by dosinabox** [dosinabox/g2_classic#99] [General]
В классике эффект был другой, надо попробовать восстановить:
![unknown](https://user-images.githubusercontent.com/39790055/195693316-4aa2bb5d-f84d-49ee-9edd-57b005cfa08d.png)
![unknown (1)](https://user-images.githubusercontent.com/39790055/195693332-7461c118-aa25-431f-a405-323e37a1e95d.png)
![unknown (2)](https://user-images.githubusercontent.com/39790055/195693351-6d6d7bfa-7ffa-49bc-93a4-fc103a14304e.png)
![unknown (3)](https://user-images.githubusercontent.com/39790055/195693363-37f4a777-b722-481c-8bc8-42eeacb220ab.png)

**Comment by dosinabox** [comment]
@Kor-Angar можешь подтвердить, что в 1.30 именно такой эффект? Пока не получается докопаться.

**Comment by Kor-Angar** [comment]
я если честно не помню. по моему где то было как на картинке у тебя. https://www.youtube.com/watch?v=epp0dMx5Vuo
у меня почему то вторая полоска в готике 2 не отображается может просто глюк.

**Comment by dosinabox** [comment]
а запиши ещё видео с телепортами и лечением из классики, пожалуйста

**Comment by dosinabox** [comment]
[о как](https://worldofplayers.ru/threads/13039/page-21#post-254361)
> в г2а мана инвестируется мгновенно, что делает недоступным ключ визуального эффекта KEY_INVEST , а вот в Г2(оригинал) мана инвестируется постепенно за какое-то время тем самым позволяя использовать KEY_INVEST

**Comment by dosinabox** [comment]
Пока отложу волну смерти, что-то не получается. Зато можно безболезненно починить эффекты телепортов и сделать их как в Г1 и классике Г2:
```cpp
func int Spell_Logic_TeleportFarm(var int manaInvested)
{
	if(manaInvested == 0)
	{
		return SPL_NEXTLEVEL;
	};
	if(manaInvested >= SPL_Cost_Teleport + 1)
	{
		return SPL_SENDCAST;
	};
	return SPL_RECEIVEINVEST;
};

func void Spell_Cast_TeleportFarm()
{
	B_PrintTeleportTooFarAway(NEWWORLD_ZEN);
	AI_Teleport(self,"BIGFARM");
};
```

**Comment by Kor-Angar** [comment]
> а запиши ещё видео с телепортами и лечением из классики, пожалуйста

вечером поставлю немецкую версию 1.30 и запишу.

>  Зато можно безболезненно починить эффекты телепортов и сделать их как в Г1 и классике Г2:

там не всё так просто. пираньи в Г2 отказались от manaInvested  для большей части заклинаний, из-за того что заклинания стали требовать много маны. а это значит, что для каждой группы заклинаний надо подбирать `time_per_mana` , то есть примерно угадать, потом проверить, потом скорректировать и окончательно утвердить.

используя старую систему для лечения и телепорта ты снизишь выживаемость героя. пока герой будет инвестировать 10 единиц маны, его орки разберут на кусочки. то же с лечением: заинвестировать такой объём маны на руне пала подразумевает что герой находится в безопасном месте. а в готики 2 уже все привыкли к "быстрому касту", то есть что можно подлечится недалеко от врага и тут же контратаковать.
к тому же Invested лечение это когда за каждую вложенную ману герой получают какое то кол-во единиц жизни и в таком случае непонятно чем будут отличать легкое лечение, от среднего и тяжелого :-) то есть они будут хуже так как за единицу маны восстанавливают меньше жизней. значит опять нужно придумывать баланс и делать что бы они лечили больше и быстрее за единицу времени.

или возможно я просто не понял что ты имеешь ввиду :-)

**Comment by Kor-Angar** [comment]
> делает недоступным ключ визуального эффекта KEY_INVEST

достаточно один раз сделать `SPL_NEXTLEVEL`. собственно вот у тебя в начале оно и идёт :-)
```D
	if(manaInvested == 0)
	{
		return SPL_NEXTLEVEL;
	};
```

**Comment by Kor-Angar** [comment]
> а вот в Г2(оригинал) мана инвестируется постепенно за какое-то время

надо смотреть про какие Spell_Logic_ идёт речь. я такого не помню. твой код для классик чистый, без правок и в игре я не встречал отличий между НВ и классик.

**Comment by Kor-Angar** [comment]
классик | НВ
-- | --
![image](https://user-images.githubusercontent.com/49514764/199873423-feb1cc3f-2c60-483d-bc46-afe566e5bfc3.png)  |  ![image](https://user-images.githubusercontent.com/49514764/199873386-8769455d-f71f-457d-af0b-bb4a15d71ab1.png)

а в НВ и Классик код одинаковый

**Comment by dosinabox** [comment]
Может, кто-нибудь сделает плагин или патч, и не придется этим заниматься:
https://worldofplayers.ru/threads/42298/post-1163067

**Comment by dosinabox** [comment]
> пока герой будет инвестировать 10 единиц маны, его орки разберут на кусочки

При `time_per_mana = 0` инвнстирование происходит мгновенно, скорость каста телепорта остается такой же. Остальные заклинания не смотрела.

**Comment by Kor-Angar** [comment]
А про что патч то? :-).  Есть заклинание контроля и телекинеза и прочее там набор анимацией. Есть анимации при 0 уровне, есть анимации при  уровне 1. Что патч будет делать? Кол-во заклинаний с некстлевел ограничено. Фактически не на что влиять не будет. Если у телепорт увеличить время инвестирования маны, ты увидишь, анимацию/эффект уровня 1, но ведь тогда будет нарушен баланс? :-) 7:38, 4 ноября 2022 г., dosinabox ***@***.***>:
Может, кто-нибудь сделает плагин или патч, и не придется этим заниматься:
https://worldofplayers.ru/threads/42298/post-1163067

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
-- Отправлено из мобильного приложения Яндекс Почты

**Comment by dosinabox** [comment]
> достаточно один раз сделать `SPL_NEXTLEVEL`. собственно вот у тебя в начале оно и идёт :-)

Помогло огненному дождю и телепорту. Волне смерти не помогло.

**Comment by dosinabox** [comment]
Можешь протестировать в бою огненный дождь с эффектом каста. Изменений в расходе маны и скорости каста быть не должно:
```cpp
func int Spell_Logic_Firerain(var int manaInvested)
{
	if(Npc_GetActiveSpellIsScroll(self) && (self.attribute[ATR_MANA] >= SPL_Cost_Scroll))
	{
		if(manaInvested == 0)
		{
			return SPL_NEXTLEVEL;
		};
		return SPL_SENDCAST;
	}
	else if(self.attribute[ATR_MANA] >= SPL_Cost_Firerain)
	{
		if(manaInvested == 0)
		{
			return SPL_NEXTLEVEL;
		};
		return SPL_SENDCAST;
	};
	return SPL_SENDSTOP;
}; 
```

**Comment by Kor-Angar** [comment]
`	if(Npc_GetActiveSpellIsScroll(self) && (self.attribute[ATR_MANA] >= SPL_Cost_Scroll))`

это код для НВ. в НВ проверять?

**Comment by Kor-Angar** [comment]
`Spell_Logic_Firerain ` нету 	`return SPL_RECEIVEINVEST;`
***

исходя из прошлого опыта оно должно выглядеть как то так

```D
func int Spell_Logic_Firerain(var int manaInvested)
{
// этот блок можны вынести в int функцию для всех подобных заклинаний что бы не повторяться.
// if (C_Spell_Logic_****(manaInvested, SPL_Cost_Firerain) == SPL_SENDSTOP)  { return SPL_SENDSTOP; };
// if (C_Spell_Logic_****(manaInvested, SPL_Cost_Firerain) == SPL_SENDCAST)  { return SPL_SENDCAST; };
	if (Npc_GetActiveSpellIsScroll(self))
	{
        if (self.attribute[ATR_MANA] < SPL_Cost_Scroll)
        && (manaInvested == 0)
        {
            return SPL_SENDSTOP;
        };
        if (manaInvested >= SPL_Cost_Scroll)
        {
            return SPL_SENDCAST;
        };
    };
    if (!Npc_GetActiveSpellIsScroll(self))
    {
        if (self.attribute[ATR_MANA] < SPL_Cost_Firerain)
        && (manaInvested == 0)
        {
            return SPL_SENDSTOP;
        };
        if (manaInvested >= SPL_Cost_Scroll)
        {
            return SPL_SENDCAST;
        };
    };
    
  	if(manaInvested == 0)
	{
        manaInvested = 1;
        return SPL_NEXTLEVEL;
	};  

	return SPL_RECEIVEINVEST;
};
```
хотя в конкретно этом случае вместо SPL_RECEIVEINVEST можно использовать и SPL_NEXTLEVEL. хотя SPL_RECEIVEINVEST  прекрасен тем что отнимает манну через движок что убирает необходимость использовать гору доп проверок, про это я писал в топике про большой огненный шар.

**Comment by dosinabox** [comment]
Усложняешь. Нам нужен только визуальный эффект при касте, не меняя механику. Для этого достаточно одного SPL_NEXTLEVEL при `manaInvested == 0`.

**Comment by Kor-Angar** [comment]
это я логическое мышление развиваю:-)  в НВ проверять? в неоф. обновление? в классик?

**Comment by dosinabox** [comment]
> в НВ проверять? в неоф. обновление? в классик?

Везде должно работать. Тогда для классики код немного другой:
```cpp
func int Spell_Logic_Firerain(var int manaInvested)
{
	if(self.attribute[ATR_MANA] >= SPL_Cost_Firerain)
	{
		if(manaInvested == 0)
		{
			return SPL_NEXTLEVEL;
		};
		return SPL_SENDCAST;
	};
	return SPL_SENDSTOP;
}; 
```

**Comment by Kor-Angar** [comment]
в мод-классик работает. да и везде будет. я сразу не понял про какой именно эффект ты писала :-)
https://www.youtube.com/watch?v=kv0jcz-aZ-E

у тебя и константа в классик оказывается есть :-) 
![image](https://user-images.githubusercontent.com/49514764/200041411-2f4bebf9-5558-4964-848e-f19cd9a73109.png)

**Comment by Kor-Angar** [comment]
> Помогло огненному дождю и телепорту. Волне смерти не помогло.

я бы увеличил для теста  `time_per_mana`. возможно там всё так стримительно что то происходит и одно на другое накладывается или что то там еще. плюс надо проверятся с фиксом и без фикса на магию от Slavemaster

**Comment by Kor-Angar** [comment]
Может быть ещё и ледяную волну посмотреть и ещё какие то заклинания типа вызова. У них тоже вполне могут буть эффекты. Ило ты уже скрипты эффектов отсмотрела? :-) 19:44, 4 ноября 2022 г., dosinabox ***@***.***>:
Усложняешь. Нам нужен только визуальный эффект при касте, не меняя механику. Для этого достаточно одного SPL_NEXTLEVEL при manaInvested == 0.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
-- Отправлено из мобильного приложения Яндекс Почты

**Comment by dosinabox** [comment]
У призывов есть эффект землетрясения при касте, починить из скриптов не получилось. У некоторых остальных есть только небольшой источник света. Самое заметное и интересное - телепорты, огненный дождь, волна смерти.
А запиши ещё сравнение лечения.

**Comment by dosinabox** [comment]
> пиши пожалуйста версии. так как мы к ветке по классик и я не много теряюсь.

Варианты:
- классика на движке 1.30 vs аддон на движке 2.6
- классика на движке 1.30 vs классика на движке 2.6
- классика на движке 1.30 vs обновление на движке 2.6

**Comment by Kor-Angar** [comment]
BorderlessWindow=1
; ... включает оконный режим без рамки, - включено (1) или выключено (0), по-умолчанию: 0.
; рекомендуется включать только при возникновении проблем с работой игры в полноэкранном режиме.


G2Classic-SystemPack-1.8
[G2Classic-SystemPack-1.8.zip](https://github.com/dosinabox/g2_classic/files/9944576/G2Classic-SystemPack-1.8.zip)

gothic2_fix-1.30.0.0
[gothic2_fix-1.30.0.0.zip](https://github.com/dosinabox/g2_classic/files/9944585/gothic2_fix-1.30.0.0.zip)

**Comment by Kor-Angar** [comment]
трансляция. секретная ссылка https://youtu.be/QlRZs0yCXFk

**Comment by dosinabox** [comment]
> достаточно один раз сделать `SPL_NEXTLEVEL`. собственно вот у тебя в начале оно и идёт :-)
> 
> ```d
> if(manaInvested == 0)
> {
> 	return SPL_NEXTLEVEL;
> };
> ```

Подходит только телепортам и огненному дождю. Было установлено, что волна смерти, армия тьмы, превращения и лечения какое-то время "заряжаются" на уровне анимации. Экспериментально было установлено, что их нужно накачивать SPL_NEXTLEVEL до достижения manaInvested в **32**. Интересно, зависит ли это число от количества кадров в секунду. Предлагаю тебе проверить у себя на движке 2.6, появятся ли эффекты:

Волна смерти:
```cpp
func int Spell_Logic_Massdeath(var int manaInvested)
{
	Print(IntToString(manaInvested));
	if(manaInvested < 32)
	{
		return SPL_NEXTLEVEL;
	};
	if(Npc_GetActiveSpellIsScroll(self) && (self.attribute[ATR_MANA] >= SPL_Cost_Scroll))
	{
		return SPL_SENDCAST;
	}
	else if(self.attribute[ATR_MANA] >= SPL_Cost_MassDeath)
	{
		return SPL_SENDCAST;
	};
	return SPL_SENDSTOP;
};
```
Армия тьмы:
```cpp
func int Spell_Logic_ArmyOfDarkness(var int manaInvested)
{
	Print(IntToString(manaInvested));
	if(manaInvested < 32)
	{
		return SPL_NEXTLEVEL;
	};
	if(Npc_GetActiveSpellIsScroll(self) && (self.attribute[ATR_MANA] >= SPL_Cost_Scroll))
	{
		return SPL_SENDCAST;
	}
	else if(self.attribute[ATR_MANA] >= SPL_Cost_ArmyOfDarkness)
	{
		return SPL_SENDCAST;
	};
	return SPL_SENDSTOP;
};
```
Превращения:
```cpp
func int Spell_Logic_TrfSheep(var int manaInvested)
{
	Print(IntToString(manaInvested));
	if(manaInvested < 32)
	{
		return SPL_NEXTLEVEL;
	};
	if(self.attribute[ATR_MANA] >= SPL_Cost_TrfSheep)
	{
		B_StartMagicTransform();
		self.attribute[ATR_MANA] -= SPL_Cost_TrfSheep;
		Npc_SetActiveSpellInfo(self,SheepTransform);
		return SPL_SENDCAST;
	};
	return SPL_SENDSTOP;
};
```
Лечение:
```cpp
func int Spell_Logic_FullHeal(var int manaInvested)
{
	Print(IntToString(manaInvested));
	if(manaInvested < 32)
	{
		return SPL_NEXTLEVEL;
	};
	if(Npc_GetActiveSpellIsScroll(self) && (self.attribute[ATR_MANA] >= SPL_Cost_Scroll))
	{
		return SPL_SENDCAST;
	}
	else if(self.attribute[ATR_MANA] >= SPL_Cost_FullHeal)
	{
		return SPL_SENDCAST;
	};
	return SPL_SENDSTOP;
};
```

**Reply by Kor-Angar** [reply]
```D
if(manaInvested == 0)
{
	return SPL_NEXTLEVEL;
};
```

> Подходит только телепортам и огненному дождю.

потому что у огненного дождя есть `MFX_FIRERAIN_INVESTGLOW` :-)

**Reply by Kor-Angar** [reply]
..

**Reply by dosinabox** [reply]
Закапываешься очень глубоко) Сначала проверь мой вариант с manaInvested < 32, пожалуйста

**Reply by Kor-Angar** [reply]
так работает
```D
func int Spell_Logic_Massdeath(var int manaInvested)
{
	Print(IntToString(manaInvested));
	if(manaInvested < 32)
	{
		return SPL_NEXTLEVEL;
	};
	if(Npc_GetActiveSpellIsScroll(self) && (self.attribute[ATR_MANA] >= SPL_Cost_Scroll))
	{
		return SPL_SENDCAST;
	}
	else if(self.attribute[ATR_MANA] >= SPL_Cost_MassDeath)
	{
		return SPL_SENDCAST;
	};
	return SPL_SENDSTOP;
};
```

**Reply by Kor-Angar** [reply]
> Закапываешься очень глубоко) Сначала проверь мой вариант с manaInvested < 32, пожалуйста

почему глубоко? я думал ты проверяла :-)  вон тебе нашёл демона ;-) и какую то старую версию "дыхания смерти" :-)

**Reply by dosinabox** [reply]
У меня работает, но я понятия не имею как это связано с числом 32 и будет ли оно одинаковым для всех.

**Reply by Kor-Angar** [reply]
армия тьмы +

**Reply by Kor-Angar** [reply]
в MDK
```D

/*INSTANCE MFX_TRANSFORM_ORIGIN (C_PARTICLEFX)
{
     ppsvalue = 100.000000000;
     ppsscalekeys_s = "1 1 1 1 2 2 3 3 4 4";
     ppsissmooth = 1;
     ppsislooping = 1;
     ppsfps = 2.000000000;
     shptype_s = "POINT";
     shpfor_s = "object";
     shpoffsetvec_s = "0 0 0";
     shpdistribtype_s = "RAND";
     shpdim_s = "10";
     shpscalekeys_s = "1";
     shpscaleislooping = 1;
     shpscaleissmooth = 1;
     shpscalefps = 2.000000000;
     dirmode_s = "RAND";
     dirfor_s = "object";
     diranglehead = 90.000000000;
     dirangleheadvar = 45.000000000;
     dirangleelev = 90.000000000;
     dirangleelevvar = 45.000000000;
     velavg = 0.600000024;
     velvar = 0.200000003;
     lsppartavg = 400.000000000;
     lsppartvar = 200.000000000;
     flygravity_s = "0 0 0";
     visname_s = "WAVEOFINSANITY_A0.TGA";
     visorientation_s = "VELO";
     vistexisquadpoly = 1;
     vistexanifps = 5.000000000;
     vistexaniislooping = 1;
     vistexcolorstart_s = "255 255 255";
     vistexcolorend_s = "255 255 255";
     vissizestart_s = "1 1";
     vissizeendscale = 1.000000000;
     visalphafunc_s = "ADD";
     visalphastart = 20.000000000;
     trlfadespeed = 0.200000003;
     trltexture_s = "LIGHTNING_BIG_A0.TGA";
     trlwidth = 3.000000000;
};*/

INSTANCE MFX_TRANSFORM_ORIGIN (C_PARTICLEFX)
{
     ppsvalue = 100;
     ppsscalekeys_s = "1 1 1 1 1";
     ppsissmooth = 1;
     ppsfps = 5;
     shptype_s = "SPHERE";
     shpfor_s = "object";
     shpoffsetvec_s = "0 0 0";
     shpdistribtype_s = "UNIFORM";
     shpdim_s = "10";
     shpmeshrender_b = 1;
     shpscalekeys_s = "1";
     shpscaleissmooth = 1;
     dirmode_s = "RAND";
     dirfor_s = "OBJECT";
     dirmodetargetfor_s = "OBJECT";
     dirmodetargetpos_s = "0 0 0";
     dirangleheadvar = 180;
     dirangleelevvar = 180;
     velavg = 0.00999999978;
     lsppartavg = 2000;
     lsppartvar = 100;
     flygravity_s = "0 0 0";
     visname_s = "HEAVENLIGHT.TGA";
     visorientation_s = "VELO3D";
     vistexanifps = 18;
     vistexaniislooping = 1;
     vistexcolorstart_s = "50 150 255";
     vistexcolorend_s = "50 0 255";
     vissizestart_s = "1 50";
     vissizeendscale = 30;
     visalphafunc_s = "ADD";
     visalphastart = 255;
     useemittersfor = 1;
};
```

**Reply by Kor-Angar** [reply]
а с лечением у тебя так же? https://www.youtube.com/watch?v=siw88nEb9t8

**Reply by Kor-Angar** [reply]
вообщем **Noki** сам себя перехитрил :-)

```D
META
{
  Parser = VFX;

};

		instance spellFX_Heal_KEY_CAST	 	(C_ParticleFXEmitKey)
		{
			// Noki: wieder aktiviert, da er an der Hand bliebt
			pfx_ppsisloopingChg = 0; // единичка была
			emCreateFXID		= "spellFX_Heal_ORIGIN";
};
```

**Reply by Kor-Angar** [reply]
жалко только что у Пала некрасивое лечение :-(

**Reply by Kor-Angar** [reply]
на первом уровне можно подключить один эффект. поэтому наверно с телепортом и дождём проблем не было так как там один эффект. сейчас попробовал на рунах Пала для смеха 

```D
instance spellFX_Heal_LEFTHAND	(CFx_Base_Proto)
{
		visname_S 			= "MFX_HEAL_CAST";
        sfxID			  	= "MFX_HEAL_CAST";
		emTrjOriginNode 	= "ZS_LEFTHAND";
		visAlpha			= 1;
	    emtrjmode_s 		= "FIXED";
	    LightPresetname		= "AURA";
};
```

и сделал простые  спелл логик

```D
func int Spell_Logic_PalLightHeal (var int manaInvested)
{
    Hlp_PrintConsole(Str_Format("manaInvested = %i", manaInvested));
	return C_Spell_Logic_SPL_SEND(manaInvested, SPL_Cost_PalLightHeal);
};

func int Spell_Logic_PalMediumHeal (var int manaInvested)
{
    Hlp_PrintConsole(Str_Format("manaInvested = %i", manaInvested));
	return C_Spell_Logic_SPL_SEND(manaInvested, SPL_Cost_PalMediumHeal);
};

func int Spell_Logic_PalFullHeal (var int manaInvested)
{
    Hlp_PrintConsole(Str_Format("manaInvested = %i", manaInvested));
    //return SPL_RECEIVEINVEST;
	return C_Spell_Logic_SPL_SEND(manaInvested, SPL_Cost_PalFullHeal);
};
```
Лёгкое и Среднее кастуется  с одним эффектом, а тяжёлое уже с двумя

[C_Spell_Logic_SPL_SEND.zip](https://github.com/dosinabox/g2_classic/files/9945267/C_Spell_Logic_SPL_SEND.zip)

**Comment by Kor-Angar** [comment]
хотел проверить эффект у заклинания. но не могу найти **PfxInstMagic**.

**Reply by dosinabox** [reply]
Это в PARTICLEFX.DAT: https://github.com/dosinabox/g2nr_unofficial_update/tree/master/PrjPARTICLEFX

**Reply by Kor-Angar** [reply]
это НВ. а я классик хотел посмтотреть

**Reply by dosinabox** [reply]
Декомпилируй себе PARTICLEFX.DAT классики. Я для мода беру этот файл из обновления, с исправлениями.

**Reply by Kor-Angar** [reply]
ясно.

**Reply by Kor-Angar** [reply]
ничего интересного там нет
[lightstar_red.zip](https://github.com/dosinabox/g2_classic/files/9944786/lightstar_red.zip)

**Comment by Kor-Angar** [comment]
а я уже вот так пробовал :-) не сработало :-D

[Spell_ArmyOfDarkness.zip](https://github.com/dosinabox/g2_classic/files/9944872/Spell_ArmyOfDarkness.zip)

**Reply by Kor-Angar** [reply]
так тоже работает
```D
    /*
  	if(manaInvested == 0)
	{
        manaInvested = 1;
        return SPL_NEXTLEVEL;
	};  
    */
    Hlp_PrintConsole(Str_Format("manaInvested = %i", manaInvested));

	return SPL_RECEIVEINVEST;
};
```

**Reply by Kor-Angar** [reply]
а не работало мой потому что я юзал свиток :-D

**Reply by dosinabox** [reply]
покажи полный код

**Reply by Kor-Angar** [reply]
вывод: Надо не забыть про свитки! я ведь даже плагины специально все поотключал которые игнорят проверку на свитки :-)

**Reply by Kor-Angar** [reply]
боюсь тебе он не понравится :-)
```D 
func int Spell_Logic_ArmyOfDarkness (var int manaInvested)
{
// этот блок можны вынести в int функцию для всех подобных заклинаний что бы не повторяться.
// if (C_Spell_Logic_****(manaInvested, SPL_Cost_Firerain) == SPL_SENDSTOP)  { return SPL_SENDSTOP; };
// if (C_Spell_Logic_****(manaInvested, SPL_Cost_Firerain) == SPL_SENDCAST)  { return SPL_SENDCAST; };
	if (Npc_GetActiveSpellIsScroll(self))
	{
        Hlp_PrintConsole("Npc_GetActiveSpellIsScroll -- true");

        if (self.attribute[ATR_MANA] < SPL_Cost_Scroll)
        && (manaInvested == 0)
        {
            return SPL_SENDSTOP;
        };
        if (manaInvested >= SPL_Cost_Scroll)
        {
            return SPL_SENDCAST;
        };
    };
    if (!Npc_GetActiveSpellIsScroll(self))
    {
        Hlp_PrintConsole("Npc_GetActiveSpellIsScroll -- false");
        if (self.attribute[ATR_MANA] < SPL_Cost_ArmyOfDarkness)
        && (manaInvested == 0)
        {
            return SPL_SENDSTOP;
        };
        if (manaInvested >= SPL_Cost_ArmyOfDarkness)
        {
            return SPL_SENDCAST;
        };
    };
    /*
  	if(manaInvested == 0)
	{
        manaInvested = 1;
        return SPL_NEXTLEVEL;
	};  
*/
    
    Hlp_PrintConsole(Str_Format("manaInvested = %i", manaInvested));
	return SPL_RECEIVEINVEST;
//	return SPL_NEXTLEVEL;
};
```

**Reply by Kor-Angar** [reply]
обновил

**Reply by Kor-Angar** [reply]
млин. а сейчас не работает.

**Reply by Kor-Angar** [reply]
- 	`return SPL_RECEIVEINVEST; `неработает
- 	`return SPL_NEXTLEVEL;` **работает**

**Comment by Kor-Angar** [comment]
картинки

**Reply by Kor-Angar** [reply]
![20221106045335_1](https://user-images.githubusercontent.com/49514764/200151998-36404a4f-7038-499d-8a21-e8ce370084fd.jpg)
![20221106045350_1](https://user-images.githubusercontent.com/49514764/200152000-172fde33-7016-4779-8564-402d034c8f07.jpg)
![20221106045436_1](https://user-images.githubusercontent.com/49514764/200152001-07c00744-3d9d-4098-a829-a5ab5f3238d8.jpg)
![20221106045447_1](https://user-images.githubusercontent.com/49514764/200152003-a8a6b1b1-7fec-4955-aa53-e94245522265.jpg)
![20221106045528_1](https://user-images.githubusercontent.com/49514764/200152005-60c6a569-5d5e-4463-88a0-91ddfc1ad76b.jpg)
![20221106045543_1](https://user-images.githubusercontent.com/49514764/200152006-8c37f7e4-8edb-45ff-be0d-d79622c3cad4.jpg)

**Comment by Kor-Angar** [comment]
```D
instance spellFX_PalHeal(CFx_Base_Proto)
{
		visname_S 			= "MFX_Heal_INIT";
		emtrjmode_s 		= "FIXED";
		emTrjOriginNode 	= "ZS_RIGHTHAND";
		//emFXInvestOrigin_S 	= "spellFX_PalHeal_ORIGIN";			// in gothic 2 sind die heal zauber instant spells, daher gibts keine investphase

		};
```

**Comment by dosinabox** [comment]
И какой вывод? Как лучше сделать?

**Reply by Kor-Angar** [reply]
вообщем я перекопал инстанции **VisualFxInst**. в НВ и классик они отличаются. и там надо порядок наводить, что бы оно работало без переключения уровней. поэтому самый простой вариант это переключать уровни как ты писала выше по твоей схеме. так как в таком случае будет решена конкретная задача - восстановление эффектов. смысла переписывать инстанции, которые и так работают, на мой взгляд никакого нет.

**Reply by Kor-Angar** [reply]
если лечение будет себя плохо вести его можно легко пофиксить
```D
META
{
  Parser = VFX;

};

		instance spellFX_Heal_KEY_CAST	 	(C_ParticleFXEmitKey)
		{
			// Noki: wieder aktiviert, da er an der Hand bliebt
			pfx_ppsisloopingChg = 0; // единичка была
			emCreateFXID		= "spellFX_Heal_ORIGIN";
};
```

**Reply by Kor-Angar** [reply]
я закинул в свой мод фикс магии более сложный вариант https://steamcommunity.com/sharedfiles/filedetails/?id=2825215672
что бы посмотреть как оно будет себя вести в игре в будущем.

**Reply by Kor-Angar** [reply]
и если думать про это в будущем, можно присмотреться к `spellFX_Geyser` это передовой опыт пираней :-)

**Reply by dosinabox** [reply]
> если лечение будет себя плохо вести его можно легко пофиксить

При `pfx_ppsisloopingChg = 0` эффект экипированного заклинания не удаляется с руки при выходе из режима магии. Эффект каста при этом завершается правильно.

**Reply by Kor-Angar** [reply]
у меня выходит. перепроверь. может просто глюк.

**Reply by Kor-Angar** [reply]
у меня в коде ещё шаблонные инстанции добавлены. не думаю конечно, что они на что то влияют... но мало ли :-)
```D
META
{
  Parser = VFX;

};
///   													XXXXXXXXXXXXXXX
///   													XX  H E A L  XX
///   													XXXXXXXXXXXXXXX
/*
instance spellFX_Heal(CFx_Base_Proto)
{
		visname_S 			= "MFX_Heal_INIT";
		emtrjmode_s 		= "FIXED";
		emTrjOriginNode 	= "ZS_RIGHTHAND";

		};
        */
		instance spellFX_Heal_KEY_OPEN(C_ParticleFXEmitKey)
		{
		};

		instance spellFX_Heal_KEY_INIT		(C_ParticleFXEmitKey)
		{
		};
        /*
		instance spellFX_Heal_KEY_INVEST_1	(C_ParticleFXEmitKey)
		{
			visname_S		= "MFX_HEAL_CAST";
			emCreateFXID	= "spellFX_Heal_LEFTHAND";
		};
        */

		instance spellFX_Heal_KEY_CAST	 	(C_ParticleFXEmitKey)
		{
			// Noki: wieder aktiviert, da er an der Hand bliebt
			pfx_ppsisloopingChg = 0; // единичка была
			emCreateFXID		= "spellFX_Heal_ORIGIN";

};
```

**Reply by dosinabox** [reply]
> у меня выходит. перепроверь. может просто глюк.

Ага, надо было в скриптах подправить

**Reply by Kor-Angar** [reply]
гуд. ты же без рендера играешь?

**Reply by dosinabox** [reply]
без

**Reply by Kor-Angar** [reply]
ну и прекрасно. значит и там и там работает :-)

**Comment by Kor-Angar** [comment]
кстати без задержки слишком рано эффект появится. возможно это как то связано с анимацией НПС. сейчас на видео покажу по кадрам

**Reply by Kor-Angar** [reply]
в одной руке спелл. руки сводятся. и спелл уже в двух руках, как бы разделившись.
https://www.youtube.com/watch?v=cXIlokYPAcM

без задержки такого эффекта не будет, так как во второй руке спел будет появляться слишком рано.

**Comment by dosinabox** [comment]
Эффекты для паладинского исцеления + исправления:
```cpp
META
{
	Parser = VFX;
};

instance SPELLFX_HEAL_KEY_CAST(C_PARTICLEFXEMITKEY)
{
	pfx_ppsisloopingchg = 0; 
	emcreatefxid = "spellFX_Heal_ORIGIN";
};

instance SPELLFX_PALHEAL_KEY_CAST(C_PARTICLEFXEMITKEY)
{
	pfx_ppsisloopingchg = 0;
	emcreatefxid = "spellFX_Heal_ORIGIN";
};

instance SPELLFX_PALHEAL_KEY_INVEST_1(C_PARTICLEFXEMITKEY)
{
	visname_s = "MFX_HEAL_CAST";
	emcreatefxid = "spellFX_Heal_LEFTHAND";
};
```

**Reply by Kor-Angar** [reply]
у них же нет этого эффекта;-)

**Reply by Kor-Angar** [reply]
инвест наверное прокатит без префикса. лучше их в логическом порядке выстроить инвест-- каст.

**Reply by dosinabox** [reply]
> у них же нет этого эффекта;-)

Вот и пусть будет.

> лучше их в логическом порядке выстроить инвест-- каст.

В скриптах так и будет. Код выше - просто патч для теста.

**Reply by Kor-Angar** [reply]
`C_ParticleFXEmitKey`

**Reply by Kor-Angar** [reply]
для скриптов :-)

**Reply by Kor-Angar** [reply]
там в  MDK кстати примечания есть

```D
instance spellFX_PalHeal_START		(CFx_Base_Proto)			// HEAL START wird im 1. Invest-Key getriggert. Sдule aus dem Boden.

instance spellFX_PalHeal_ORIGIN	(CFx_Base_Proto)		// HEAL ORIGIN wird automatisch ьber emFXInvestOrigin_S getriggert. Aura um den Spieler
```

**Comment by Kor-Angar** [comment]
посмотрел. у меня в порте Готики 1 еще используется `SPL_STATUS_CANINVEST_NO_MANADEC`

```D
    return SPL_STATUS_CANINVEST_NO_MANADEC; // неотнимает ману
//  return SPL_RECEIVEINVEST;               // отнимает ману
```
